openssl/src/cms.rs

```diff
@@ -236,4 +313,61 @@ mod test {
 
         assert_eq!(input, decrypt);
     }
+
+    #[test]
+    fn cms_sign_verify() {
+        // load cert with private key
+        let priv_cert_bytes = include_bytes!("../test/cms.p12");
+        let priv_cert = Pkcs12::from_der(priv_cert_bytes).expect("failed to load priv cert");
+        let priv_cert = priv_cert.parse("mypass").expect("failed to parse priv cert");
+
+        // sign cms message using private key cert
+        let input = String::from("My Message");
+        let sign = CmsContentInfo::sign(Some(&priv_cert.cert), Some(&priv_cert.pkey), None, Some(&input.as_bytes()), CMSOptions::empty())
+            .expect("failed create signed cms");
+        let sign = sign.to_der().expect("failed to create der from cms");
+
+        // verify signature on cms message
+        let verify = CmsContentInfo::from_der(&sign).expect("failed read cms from der");
+
+        let mut cert_stack = Stack::new().expect("failed to create stack");
+        cert_stack.push(priv_cert.cert.clone()).expect("failed to add cert to stack");
+
+        let mut store_builder = X509StoreBuilder::new().expect("failed to create store builder");
+        store_builder.add_cert(priv_cert.cert.clone()).expect("failed to add certificate to store");
+        let store = store_builder.build();
+
+        let verify = verify.verify(Some(&cert_stack), &store, CMSOptions::empty()).expect("failed to verify cms");
+        let verify = verify.expect("cms verification returned None");
+        let verify = String::from_utf8(verify).expect("failed to create string from cms content");
+
+        assert_eq!(input, verify);
+    }
+
+    #[test]
+    fn cms_sign_verify_detached() {
+        // load cert with private key
+        let priv_cert_bytes = include_bytes!("../test/cms.p12");
+        let priv_cert = Pkcs12::from_der(priv_cert_bytes).expect("failed to load priv cert");
+        let priv_cert = priv_cert.parse("mypass").expect("failed to parse priv cert");
+
+        // sign cms message using private key cert
+        let input = String::from("My Message");
+        let sign = CmsContentInfo::sign(Some(&priv_cert.cert), Some(&priv_cert.pkey), None, Some(&input.as_bytes()), CMSOptions::DETACHED)
```

Comment:
It seems like the sign API should match the verify API one way or another. Either separate methods for attached and detached signatures, or one combined for both.

Reply:
Does it make sense to make `sign()` function which just adds extra flag? Unlike `verify()`, detached `sign()` needs exactly the same input and has exactly the same output type.

Regarding the function signature of `sign()`, the first and the last parameters are always `Some`, so one probably should change them to just values instead of `Option`. This would be a 
breaking change.

BTW, the AppVeyor Windows test seems to fail at the initialization phase before actual checks, isn't it? Or does this PR has some issues with tests?

Reply:
> Regarding the function signature of `sign()`, the first and the last parameters are always `Some`, so one probably should change them to just values instead of `Option`. This would be a
> breaking change.

That doesn't seem to be the case:

> Any or all of these parameters can be NULL, see NOTES below.

> If signcert and pkey are NULL then a certificates only CMS structure is output.

https://www.openssl.org/docs/man1.1.0/man3/CMS_sign.html

> BTW, the AppVeyor Windows test seems to fail at the initialization phase before actual checks, isn't it? Or does this PR has some issues with tests?

That's not caused by this PR - vcpkg's openssl build is broken for some reason.

Reply:
The certificates only CMS structure seems to be a special case which is not the expected output of normal usage of `sign()`. I believe it should be a separate function. This mode requires both `signcert` and `pkey` to be NULL at the same time, so it's more like an `Option<(signcert, pkey)>` than `Option<signcert>, Option<pkey>`. (In fact, I'd rather prefer `CmsContentInfo` to be an enum, but this seems to be not doable with openssl).

Reply:
Are there any actions planned on this pull request?

Reply:
Just so I understand, we've gone from "the first and last parameters are always Some" to "well, they're only None in a special case", right?

My preference is generally to expose OpenSSL APIs as directly as is possible because any single person is probably not going to understand the totality of all possible usage of them.

Reply:
Does this mean there should be another layer of abstraction on top of openssl crate?
--